client/client-multi/client-c/src/kaa/collections/kaa_list.c

```diff
@@ -309,3 +309,17 @@ void kaa_list_sort(kaa_list_t *list, match_predicate pred)
     }
     list->tail = node;
 }
+
+int32_t kaa_list_hash(kaa_list_t *list, list_node_hash pred)
+{
+    KAA_RETURN_IF_NIL2(list, pred, 0);
+    int32_t result = 1;
+    kaa_list_node_t *node = kaa_list_begin(list);
+    uint64_t element;
+    while (node) {
+        element = pred(node->data);
+        result = 31 * result + (uint32_t) (element ^ (element >> 32));
```

Comment:
Java integer overflow behaviour for this case are binary equal to C uint, because for Java it wraps around (Integer.MAX_VALUE +1 == Integer.MIN_VALUE) and for unit32_t (0x7fffffff+1 == 0x80000000). So we estimate it in uint32 but convert to int32 to be compatible with other clients.